### PR TITLE
Martin/metal updates

### PR DESCRIPTION
**Pull request by MartinJibber**
https://user-images.githubusercontent.com/88675954/165649105-1fcaedc2-1243-4f0d-b792-6aa902403aa2.MP4

### CODE REVIEW

File: Shared/Extensions/UI Extensions/UIImage+Extensions.swift [hunk 1]
```diff
@@ -35,22 +35,28 @@ extension UIImage {
         return image!
     }
 
-    func scaled(by scale: CGFloat) -> UIImage {
-        let size = self.size.applying(CGAffineTransform(scaleX: scale, y: scale))
-        let hasAlpha = false
-        let scale: CGFloat = 0.0 // Automatically use scale factor of main screen
+    /// Returns a new version of this image scaled proportionally so no side is greater than maxSideLength.
+    func imageWith(maxSideLength: CGFloat) -> UIImage {
```
Comment (MartinJibber) [inline Shared/Extensions/UI Extensions/UIImage+Extensions.swift:39]: We can't compress the image any more, but we can make a new smaller image .

File: Shared/Flows/Face/FaceDisclosureViewController.swift [hunk 2]
```diff
@@ -84,7 +84,7 @@ class FaceDisclosureViewController: DisclosureModalViewController {
 
         switch self.captureType {
         case .smiling:
-            let file = PFFileObject(name:"small_image.jpeg", data: data)
+            let file = PFFileObject(name:"small_image.png", data: data)
```
Comment (MartinJibber) [inline Shared/Flows/Face/FaceDisclosureViewController.swift:87]: Reflecting new image type.

File: Shared/UI/Face Detection/FaceDetectionViewController.swift [hunk 3]
```diff
@@ -22,29 +22,23 @@ class FaceDetectionViewController: ImageCaptureViewController {
     @Published var eyesAreClosed = false
     @Published var isSmiling = false
 
-    // The Metal pipeline.
-    var metalDevice: MTLDevice!
-    var metalCommandQueue: MTLCommandQueue!
-
-    // The Core Image pipeline.
-    var ciContext: CIContext!
```
Comment (MartinJibber) [inline Shared/UI/Face Detection/FaceDetectionViewController.swift:30]: Moved into the new metal view.

File: Shared/UI/Face Detection/FaceDetectionViewController.swift [hunk 4]
```diff
@@ -92,31 +88,128 @@ class FaceDetectionViewController: ImageCaptureViewController {
...
+            self.isSmiling = false
         }
 
+        // CGImages play nicer with UIKit.
+        // Per the docs: "Due to Core Image's coordinate system mismatch with UIKit, this filtering
+        // approach may yield unexpected results when displayed in a UIImageView with contentMode."
+        let context = CIContext()
+        let cgImage = context.createCGImage(finalCIImage, from: finalCIImage.extent)!
 
+        let image = UIImage(cgImage: cgImage, scale: 1, orientation: .up)
```
Comment (MartinJibber) [inline Shared/UI/Face Detection/FaceDetectionViewController.swift:132]: It's annoying, but ImageViews have trouble displaying CI images properly. Converting to a CG image.

File: Shared/UI/Face Detection/FaceDetectionViewController.swift [hunk 5]
```diff
@@ -92,31 +88,129 @@ class FaceDetectionViewController: ImageCaptureViewController {
...
             self.eyesAreClosed = face.leftEyeClosed && face.rightEyeClosed
             self.isSmiling = face.hasSmile
+
+            // Increase the bounds around the face so it's not too zoomed in.
+            var adjustedFaceBounds = face.bounds
+            adjustedFaceBounds.size.height = face.bounds.height * 1.8
+            adjustedFaceBounds.size.width = adjustedFaceBounds.height
+            adjustedFaceBounds.centerY = face.bounds.centerY + face.bounds.height * 0.2
+            adjustedFaceBounds.centerX = face.bounds.centerX
+
```
Comment (MartinJibber) [inline Shared/UI/Face Detection/FaceDetectionViewController.swift:119]: No two faces are the same, and the angle you take the picture can make a big difference so this is a best guess at giving the face some extra room.

We can revisit this later, but it works decently at the moment.

File: Shared/UI/Face Detection/FaceDetectionViewController.swift [hunk 6]
```diff
@@ -92,31 +88,129 @@ class FaceDetectionViewController: ImageCaptureViewController {
...
+        let cgImage = context.createCGImage(finalCIImage, from: finalCIImage.extent)!
 
+        let image = UIImage(cgImage: cgImage, scale: 1, orientation: .up)
         self.didCapturePhoto?(image)
     }
+
+    // MARK: - Process Results
+
+    // Performs the blend operation.
+    func blend(original framePixelBuffer: CVPixelBuffer, mask maskPixelBuffer: CVPixelBuffer) {
```
Comment (MartinJibber) [inline Shared/UI/Face Detection/FaceDetectionViewController.swift:139]: Just moved this all into the same file.

File: Shared/UI/Face Detection/FaceDetectionViewController.swift (hunk 5 above)
Comment (Benjamindodgson) [reply Shared/UI/Face Detection/FaceDetectionViewController.swift:119]: sick

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**